File: include/PolyWorksSDK/samples/applications/IMInspect/VisualBasic.Net/VBNetReadAndFit/MainForm.vb
Public Class MainForm

    ' Global COM Message Filter to prevent COM calls from being rejected if the server application is temporarily busy
    Private comFilter As CComMessageFilter = Nothing

    ' Global access to a PolyWorks|Inspector
    Private imInspectObject As IMInspectLib.IMInspect = Nothing

    ' Upon loading the Form
    Private Sub MainForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        comFilter = New CComMessageFilter

        imInspectObject = New IMInspectLib.IMInspect

    End Sub

    ' Cleanup upon destruction of the form
    Private Sub MainForm_FormClosed(ByVal sender As System.Object, ByVal e As System.Windows.Forms.FormClosedEventArgs) Handles MyBase.FormClosed

        System.Runtime.InteropServices.Marshal.ReleaseComObject(imInspectObject)
        imInspectObject = Nothing

        comFilter = Nothing

    End Sub

    ' Main action of this application
    Private Sub ButtonSend_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ButtonSend.Click

        ResultFittedSphere.Text = ""
        ResultAllSelectedSphere.Text = ""
        LabelResultAllSelectedSphere.Text = "Name of all spheres:"

        ' If the server does not exists, return
        If imInspectObject Is Nothing Then

            Exit Sub

        End If

        ' Obtain PolyWorks|Inspector's current project
        Dim projectObject As IMInspectLib.IIMInspectProject = Nothing
        imInspectObject.ProjectGetCurrent(projectObject)

        If projectObject Is Nothing Then

            Exit Sub

        End If

        ' Read the file and fill the coordinates and normals lists
        Dim coordinates As System.Collections.Generic.List(Of Single) = New System.Collections.Generic.List(Of Single)()
        Dim normals As System.Collections.Generic.List(Of Single) = New System.Collections.Generic.List(Of Single)()
        Dim colors As System.Collections.Generic.List(Of Byte) = New System.Collections.Generic.List(Of Byte)()

        If FillArraysFromFile((TextFileName.Text), coordinates, normals) = False Then

            Exit Sub

        End If


        ' Set all points to a blue color
        Dim index As Integer
        index = 0

        While index < coordinates.Count

            colors.Add(32)
            colors.Add(32)
            colors.Add(224)
            index += 3

        End While

        ' Obtain the name of the file, without the path
        Dim tokens As String() = TextFileName.Text.Split(New [Char]() {"\"c})
        Dim objectName As String = tokens(tokens.Length - 1)

        ' Create a point cloud in the current PolyWorks|Inspector project
        Dim pointCloudObject As IMInspectLib.IIMPointCloud
        projectObject.PointCloudCreateData(pointCloudObject, objectName, 0)

        ' Send the points to the point cloud
        If Not pointCloudObject Is Nothing Then

            pointCloudObject.PointsAdd(coordinates.ToArray(), normals.ToArray(), colors.ToArray())

        End If

        ' Obtain a Command Center from the current PolyWorks|Inspector project
        Dim commandCenterObject As IMInspectLib.IIMCommandCenter
        projectObject.CommandCenterCreate(commandCenterObject)

        ' Send commands to fit a sphere on all the elements of this new object
        Dim returnValue As Integer
        Dim script As String
        Dim sphereName As String

        If Not commandCenterObject Is Nothing Then

            commandCenterObject.CommandExecute("TREEVIEW OBJECT SELECT NONE")

            ' Perform the fit
            commandCenterObject.CommandExecute("TREEVIEW OBJECT_ELEMENTS SELECT ALL ( """ & objectName & """ )")
            commandCenterObject.CommandExecute("FEATURE PRIMITIVE SPHERE FIT_FROM_SELECTED_ELEMENTS")
            If commandCenterObject.ReturnValueIsSuccess(returnValue) = 0 Then

                ResultFittedSphere.Text = "The fitting was unsuccessful"
                Exit Sub

            End If

            ' Gets the name of the last fitted sphere. After the fitting, it is the only object selected.
            Dim lastSphereName As String
            If Not ValueStrGetFromApp(commandCenterObject, "TREEVIEW OBJECT PROPERTIES NAME GET", lastSphereName, False, "") Then

                ResultFittedSphere.Text = "Unable to obtain the name of the fitted sphere"
            Else
                ' Feedback of the name of the last fitted sphere
                ResultFittedSphere.Text = "Sphere """ & lastSphereName & """ was successfully fitted"
            End If

            ' Gets the name of all spheres
            commandCenterObject.CommandExecute("TREEVIEW OBJECT SELECT NONE")
            commandCenterObject.CommandExecute("TREEVIEW PRIMITIVE SPHERE SELECT ALL")

            Dim sphereNames As System.Collections.Generic.List(Of System.Collections.Generic.List(Of String)) = New System.Collections.Generic.List(Of System.Collections.Generic.List(Of String))
            Dim res As Boolean = ValueStrArrayGetFromApp(commandCenterObject, "TREEVIEW OBJECT NAME SELECTED GET", sphereNames, False, "")

            ' Feedback of the name of all spheres
            If (res And (Not (sphereNames.Count = 0))) Then
                NameSpheresFeedback(sphereNames(0), LabelResultAllSelectedSphere, ResultAllSelectedSphere)
            Else
                ResultAllSelectedSphere.Text = "Unable to obtain the name of the selected objects"
            End If

        End If

        ' Cleanup
        System.Runtime.InteropServices.Marshal.ReleaseComObject(pointCloudObject)
        System.Runtime.InteropServices.Marshal.ReleaseComObject(commandCenterObject)
        System.Runtime.InteropServices.Marshal.ReleaseComObject(projectObject)

        pointCloudObject = Nothing
        commandCenterObject = Nothing
        projectObject = Nothing

    End Sub

    '=================================================================================================
    ' Indicates if the specified list of strings Is empty.
    '
    ' Parameter : sphereNames_ : sphere names
    '
    ' Return : true if empty, false otherwise
    '=================================================================================================
    Private Function IsSphereNamesEmpty(ByRef sphereNames_ As System.Collections.Generic.List(Of String)) As Boolean

        If (sphereNames_.Count = 0) Then Return True

        ' Note that in MSCL, an empty array has a size of 1 And that the corresponding element Is always 0 whatever its type (int, float, string...).
        If (sphereNames_.Count = 1) Then
            If (sphereNames_(0) = "0") Then
                Return True
            End If
        End If

        Return False
    End Function

    '=================================================================================================
    ' Feedback of the name of all spheres
    '
    ' Parameter : sphereNames_                  : spher names
    ' Parameter : labelResultAllSelectedSphere_ : (out) label result to feedback
    ' Parameter : resultAllSelectedSphere_      : (out) result to feedback
    '=================================================================================================
    Private Function NameSpheresFeedback(ByRef sphereNames_ As System.Collections.Generic.List(Of String),
                                         ByRef labelResultAllSelectedSphere_ As Label,
                                         ByRef resultAllSelectedSphere_ As Label)

        Dim suffix As String = "sphere"

        Dim nbSpheres As Integer = sphereNames_.Count
        If nbSpheres = 1 Then
            If sphereNames_(0).Equals("0") Then
                nbSpheres = 0
            End If
        End If

        If nbSpheres > 1 Then
            suffix = "spheres"
        End If

        labelResultAllSelectedSphere_.Text = "Name of all spheres: ( " & nbSpheres & " " & suffix & " )"

        For i As Integer = 0 To nbSpheres - 1
            If i > 0 Then
                resultAllSelectedSphere_.Text += ", "
            End If
            resultAllSelectedSphere_.Text += sphereNames_(i)
        Next
    End Function

    '=================================================================================================
    '   Retrieves a string value from the application, using a command.
    '
    ' Parameter : commandCenter_ : command center
    ' Parameter : cmdName_       : command name
    ' Parameter : value_         : (out) retrieved value
    ' Parameter : withEcho_      : indicates whether the command should be visible in the application's Command History
    ' Parameter : cmdArg_        : command arguments
    '
    ' Return : true if successful, false otherwise
    ' -------------------------------------------------------------------------------------------------
    Private Function ValueStrGetFromApp(
        ByRef commandCenter_ As IMInspectLib.IIMCommandCenter,
        ByRef cmdName_ As String,
        ByRef value_ As String,
        ByRef withEcho_ As Boolean,
        ByRef cmdArg_ As String) As Boolean

        Dim values As System.Collections.Generic.List(Of System.Collections.Generic.List(Of String)) = New System.Collections.Generic.List(Of System.Collections.Generic.List(Of String))

        If Not ValueStrArrayGetFromApp(commandCenter_, cmdName_, values, withEcho_, cmdArg_) Then
            Return False
        End If

        If ((values.Count = 0) OrElse
            (values(0).Count = 0)) Then
            Return False
        End If

        value_ = values(0)(0)

        Return True

    End Function

    '=================================================================================================
    '   Retrieves an array of string values from the application, using a command.
    '
    ' Parameter : commandCenter_ : command center
    ' Parameter : cmdName_       : command name
    ' Parameter : values_        : (out) retrieved values
    ' Parameter : withEcho_      : indicates whether the command should be visible in the application's Command History
    ' Parameter : cmdArg_        : command arguments
    '
    ' Return : true if successful, false otherwise
    ' -------------------------------------------------------------------------------------------------
    Private Function ValueStrArrayGetFromApp(
        ByRef commandCenter_ As IMInspectLib.IIMCommandCenter,
        ByRef cmdName_ As String,
        ByRef values_ As System.Collections.Generic.List(Of System.Collections.Generic.List(Of String)),
        ByRef withEcho_ As Boolean,
        ByRef cmdArg_ As String) As Boolean

        Dim valueVar As String = "value"
        Dim script As String = "version ""5.0""" & vbCrLf &
                               "DECLARE " & valueVar & vbCrLf &
                                cmdName_ & " ( "
        If (Not String.IsNullOrEmpty(cmdArg_)) Then
            script += """" + cmdArg_ + """, "
        End If
        script += valueVar & " )"

        Dim values As System.Collections.Generic.List(Of String) = New System.Collections.Generic.List(Of String)()

        Dim valueVarArray As System.Collections.Generic.List(Of String) = New System.Collections.Generic.List(Of String)()
        valueVarArray.Add(valueVar)

        If Not RunScriptAndReadValueArrays(commandCenter_, script, valueVarArray, values_, withEcho_) Then
            Return False
        End If

        Return True

    End Function

    '=================================================================================================
    '   Retrieves the values of each variable when running a given script
    '
    ' Parameter : commandCenter_ : command center
    ' Parameter : script_        : script to execute
    ' Parameter : varNames_      : name of the variables to retrieve
    ' Parameter : values_        : retrieved values of each variable
    ' Parameter : withEcho_      : indicates whether the command should be visible in the application's Command History
    '
    ' Return : true if successful, false otherwise
    ' -------------------------------------------------------------------------------------------------
    Private Function RunScriptAndReadValueArrays(
        ByRef commandCenter_ As IMInspectLib.IIMCommandCenter,
        ByRef script_ As String,
        ByRef varNames_ As System.Collections.Generic.List(Of String),
        ByRef values_ As System.Collections.Generic.List(Of System.Collections.Generic.List(Of String)),
        ByRef withEcho_ As Boolean) As Boolean

        If (withEcho_ = True) Then
            commandCenter_.CommandExecute("EchoOn")
            commandCenter_.CommandExecute("RecordableOn")
        Else
            commandCenter_.CommandExecute("EchoOff")
            commandCenter_.CommandExecute("RecordableOff")
        End If

        ' Run the script
        Dim returnValue = commandCenter_.ScriptExecuteFromBuffer(script_, "")
        If commandCenter_.ReturnValueIsSuccess(returnValue) = 0 Then
            Return False
        End If

        ' Retrieves the values of each variable
        For i As Integer = 0 To varNames_.Count - 1

            Dim nbValues As String
            Dim values As System.Collections.Generic.List(Of String) = New System.Collections.Generic.List(Of String)

            nbValues = commandCenter_.ScriptVariableGetNbValues(varNames_(i))

            For j As Integer = 0 To nbValues - 1

                Dim value As String
                returnValue = commandCenter_.ScriptVariableGetValueAsString(varNames_(i), j + 1, value)

                If commandCenter_.ReturnValueIsSuccess(returnValue) = 1 Then
                    values.Add(value)
                End If
            Next
            values_.Add(values)

        Next

        Return True

    End Function

    '=================================================================================================
    ' Reads an ASCII file containing lines made of 6 values (X,Y,Z,I,J,K)
    ' and fills a coordinates and a normals arrays with those values
    '=================================================================================================
    Private Function FillArraysFromFile(
        ByRef filename_ As String,
        ByRef coordinates_ As System.Collections.Generic.List(Of Single),
        ByRef normals_ As System.Collections.Generic.List(Of Single)) As Boolean

        ' Validate file name
        If Not System.IO.File.Exists(filename_) Then

            Return False

        End If

        ' Open the file for reading
        Dim reader As New System.IO.StreamReader(filename_)
        Dim line As String
        Dim success As Boolean

        ' Read the first six tokens (X,Y,Z,I,J,K) on each line of the file
        ' and convert them to double-precision floating point values
        success = True
        line = reader.ReadLine()

        While Not line Is Nothing

            Dim tokens As String() = line.Split(New [Char]() {" "c})

            If tokens.Length < 6 Then

                success = False
                Exit While

            End If

            ' Recording coordinates
            coordinates_.Add(Convert.ToSingle(tokens(0)))
            coordinates_.Add(Convert.ToSingle(tokens(1)))
            coordinates_.Add(Convert.ToSingle(tokens(2)))

            ' Recording normals
            normals_.Add(Convert.ToSingle(tokens(3)))
            normals_.Add(Convert.ToSingle(tokens(4)))
            normals_.Add(Convert.ToSingle(tokens(5)))

            line = reader.ReadLine()

        End While

        reader.Close()

        Return success

    End Function

End Class

File: include/PolyWorksSDK/samples/applications/IMInspect/CSharp/CSharpReadAndFit/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.IO;
using System.Text;
using System.Windows.Forms;

namespace CSharpReadAndFit
{
    public partial class MainForm : Form
    {
        // =========================================================================================
        // ========================================= Members =======================================

        private IMInspectLib.IMInspect m_imInspect = null;

        // =========================================================================================
        // ========================================= Methods =======================================

        // =========================================================================================
        // Constructor
        // -----------------------------------------------------------------------------------------
        public MainForm()
        {
            InitializeComponent();
        }

        // =========================================================================================
        // Upon loading the form
        // -----------------------------------------------------------------------------------------
        private void MainForm_Load(object sender, EventArgs e)
        {
            m_imInspect = new IMInspectLib.IMInspect();
        }

        // =========================================================================================
        // Cleanup upon destruction of the form
        // -----------------------------------------------------------------------------------------
        private void MainForm_FormClosed(object sender, FormClosedEventArgs e)
        {
            System.Runtime.InteropServices.Marshal.ReleaseComObject(m_imInspect);
            m_imInspect = null;
        }

        // =========================================================================================
        // Main action of this application
        // -----------------------------------------------------------------------------------------
        private void ButtonSend_Click(object sender, EventArgs e)
        {
            ResultFittedSphere.Text = "";
            ResultAllSelectedSphere.Text = "";
            LabelResultAllSelectedSphere.Text = "Name of all spheres:";

            // If the server does not exists, return
            if (m_imInspect == null)
            {
                return;
            }

            // Obtain PolyWorks|Inspector's current project
            IMInspectLib.IIMInspectProject inspectProject = null;
            m_imInspect.ProjectGetCurrent(out inspectProject);

            if (inspectProject == null)
            {
                return;
            }

            // Read the file and fill the coordinates and normals lists
            System.Collections.Generic.List<float> coordinates = new System.Collections.Generic.List<float>();
            System.Collections.Generic.List<float> normals = new System.Collections.Generic.List<float>();
            System.Collections.Generic.List<Byte> colors = new System.Collections.Generic.List<Byte>();

            if (!FillArraysFromFile(TextFileName.Text, ref coordinates, ref normals))
            {
                return;
            }

            // Set all points to a blue color
            for (int index = 0; index < coordinates.Count; index += 3)
            {
                colors.Add(32);
                colors.Add(32);
                colors.Add(224);
            }

            // Obtain the name of the file, without the path
            string[] tokens = TextFileName.Text.Split(new Char[] { '\\' });
            string objectName = tokens[tokens.Length - 1];

            // Create a point cloud in the current PolyWorks|Inspector project
            IMInspectLib.IIMPointCloud pointCloud = null;
            inspectProject.PointCloudCreateData(out pointCloud, objectName, 0);

            // Send the points to the point cloud
            if (pointCloud != null)
            {
                Array coordinatesArray = coordinates.ToArray();
                Array normalsArray = normals.ToArray();
                Array colorsArray = colors.ToArray();

                pointCloud.PointsAdd(ref coordinatesArray, ref normalsArray, ref colorsArray);
            }

            // Obtain a Command Center from the current PolyWorks|Inspector project
            IMInspectLib.IIMCommandCenter commandCenter = null;
            inspectProject.CommandCenterCreate(out commandCenter);

            // Send commands to fit a sphere on all the elements of this new object
            int returnValue = 0;

            if (commandCenter != null)
            {
                commandCenter.CommandExecute("TREEVIEW OBJECT SELECT NONE");

                // Perform the fit
                commandCenter.CommandExecute("TREEVIEW OBJECT_ELEMENTS SELECT ALL ( \"" + objectName + "\" )");
                returnValue = commandCenter.CommandExecute("FEATURE PRIMITIVE SPHERE FIT_FROM_SELECTED_ELEMENTS");
                if (commandCenter.ReturnValueIsSuccess(returnValue) == 0)
                {
                    ResultFittedSphere.Text = "The fitting was unsuccessful";
                }

                // Gets the name of the last fitted sphere. After the fitting, it is the only object selected.
                String lastSphereName = "";
                if (!ValueStrGetFromApp(commandCenter, "TREEVIEW OBJECT PROPERTIES NAME GET", ref lastSphereName, false, ""))
                {
                    ResultFittedSphere.Text = "Unable to obtain the name of the fitted sphere";
                }
                else
                {
                    // Feedback of the name of the last fitted sphere
                    ResultFittedSphere.Text = "Sphere " + lastSphereName + " was successfully fitted";
                }

                // Gets the name of all spheres
                commandCenter.CommandExecute("TREEVIEW OBJECT SELECT NONE");
                commandCenter.CommandExecute("TREEVIEW PRIMITIVE SPHERE SELECT ALL");

                System.Collections.Generic.List<System.Collections.Generic.List<String>> sphereNames = new System.Collections.Generic.List<System.Collections.Generic.List<String>>();
                bool res = ValueStrArrayGetFromApp(commandCenter, "TREEVIEW OBJECT NAME SELECTED GET", ref sphereNames, false, "");

                // Feedback of the name of all spheres
                if (res && (sphereNames.Count != 0))
                {
                    NameSpheresFeedback(sphereNames[0], ref LabelResultAllSelectedSphere, ref ResultAllSelectedSphere);
                }
                else
                {
                    ResultAllSelectedSphere.Text = "Unable to obtain the name of the selected objects";
                }
            }

            // Cleanup
            System.Runtime.InteropServices.Marshal.ReleaseComObject(pointCloud);
            pointCloud = null;
            System.Runtime.InteropServices.Marshal.ReleaseComObject(commandCenter);
            commandCenter = null;
            System.Runtime.InteropServices.Marshal.ReleaseComObject(inspectProject);
            inspectProject = null;
        }

        // =========================================================================================
        // ======================================== Utilities ======================================

        //=================================================================================================
        // Indicates if the specified list of strings is empty.
        //
        // Parameter : sphereNames_ : sphere names
        //
        // Return : true if empty, false otherwise
        //=================================================================================================
        private bool IsSphereNamesEmpty(System.Collections.Generic.List<String> sphereNames_)
        {
            if (sphereNames_.Count == 0) return true;

            // Note that in MSCL, an empty array has a size of 1 and that the corresponding element is always 0 whatever its type (int, float, string...).
            if (sphereNames_.Count == 1)
            {
                if (sphereNames_[0].Equals("0"))
                    return true;
            }

            return false;
        }

        //=================================================================================================
        // Feedback of the name of all spheres
        //
        // Parameter : sphereNames_                  : sphere names
        // Parameter : labelResultAllSelectedSphere_ : (out) label result to feedback
        // Parameter : resultAllSelectedSphere_      : (out) result to feedback
        //=================================================================================================
        private void NameSpheresFeedback(System.Collections.Generic.List<String> sphereNames_,
                                         ref Label labelResultAllSelectedSphere_,
                                         ref Label resultAllSelectedSphere_)
        {
            String suffix = "sphere";
            int nbSpheres = 0;

            if ( !IsSphereNamesEmpty(sphereNames_) )
            {
                nbSpheres = sphereNames_.Count;
                if (nbSpheres > 1) suffix = "spheres";
            }

            labelResultAllSelectedSphere_.Text = "Name of all spheres: ( " + nbSpheres + " " + suffix + " )";

            for( int i = 0; i < nbSpheres; i++ )
            {
                if (i > 0) resultAllSelectedSphere_.Text += ", ";

                resultAllSelectedSphere_.Text += sphereNames_[i];
            }
        }

        // =========================================================================================
        // Reads an ASCII file containing lines made of 6 values (X,Y,Z,I,J,K)
        // and fills a coordinates and a normals lists with those values
        // -----------------------------------------------------------------------------------------
        private bool FillArraysFromFile(
            String filename_,
            ref System.Collections.Generic.List<float> coordinates_,
            ref System.Collections.Generic.List<float> normals_)
        {
            // Validate file name
            if ( !File.Exists( filename_ ) )
            {
                return false;
            }

            // Open the file for reading
            StreamReader reader = new StreamReader( filename_ );
            String line;
            bool success = true;

            // Read the first six tokens (X,Y,Z,I,J,K) on each line of the file
            // and convert them to double-precision floating point values
            while ( ( line = reader.ReadLine() ) != null )
            {
                string[] tokens = line.Split( new Char[] { ' ' } );

                if ( tokens.Length < 6 )
                {
                    success = false;

                    break;
                }

                // Recording coordinates
                coordinates_.Add( Convert.ToSingle( tokens[ 0 ] ) );
                coordinates_.Add( Convert.ToSingle( tokens[ 1 ] ) );
                coordinates_.Add( Convert.ToSingle( tokens[ 2 ] ) );

                // Recording normals
                normals_.Add( Convert.ToSingle( tokens[ 3 ] ) );
                normals_.Add( Convert.ToSingle( tokens[ 4 ] ) );
                normals_.Add( Convert.ToSingle( tokens[ 5 ] ) );
            }

            reader.Close();

            return success;
        }

        //=================================================================================================
        //   Retrieves a string value from the application, using a command.
        //
        // Parameter : commandCenter_ : command center
        // Parameter : cmdName_       : command name
        // Parameter : value_         : (out) retrieved value
        // Parameter : withEcho_      : indicates whether the command should be visible in the application's Command History
        // Parameter : cmdArg_        : command arguments
        //
        // Return : true if successful, false otherwise
        // -------------------------------------------------------------------------------------------------
        private bool ValueStrGetFromApp(
            IMInspectLib.IIMCommandCenter commandCenter_,
            String cmdName_,
            ref String value_,
            bool withEcho_,
            String cmdArg_)
        {
            value_ = "";
            System.Collections.Generic.List<System.Collections.Generic.List<String>> values = new System.Collections.Generic.List<System.Collections.Generic.List<String>>();

            if (!ValueStrArrayGetFromApp(commandCenter_, cmdName_, ref values, withEcho_, cmdArg_))
            {
                return false;
            }

            if( (values.Count == 0) || (values[0].Count == 0) )
            {
                return false;
            }

            value_ = values[0][0];
            return true;
        }

        // =================================================================================================
        //    Retrieves an array of string values from the application, using a command.
        //
        //  Parameter : commandCenter_ : command center
        //  Parameter : cmdName_       : command name
        //  Parameter : values_        : (out) retrieved values
        //  Parameter : withEcho_      : indicates whether the command should be visible in the application's Command History
        //  Parameter : cmdArg_        : command arguments
        //
        //  Return : true if successful, false otherwise
        //  -------------------------------------------------------------------------------------------------
        private bool ValueStrArrayGetFromApp(
            IMInspectLib.IIMCommandCenter commandCenter_,
            String cmdName_,
            ref System.Collections.Generic.List<System.Collections.Generic.List<String>> values_,
            bool withEcho_,
            String cmdArg_)
        {
            String valueVar = "value";
            String script = "version \"5.0\"" + System.Environment.NewLine;
            script += "DECLARE " + valueVar + System.Environment.NewLine;
            script += cmdName_ + " ( ";
            if (!String.IsNullOrEmpty(cmdArg_))
            {
                script += "\"" + cmdArg_ + "\", ";
            }
            script += valueVar + " )";

            System.Collections.Generic.List<String> values = new System.Collections.Generic.List<String>();
            System.Collections.Generic.List<String> valueVarArray = new System.Collections.Generic.List<String>();
            valueVarArray.Add(valueVar);
            if ( ! RunScriptAndReadValueArrays(commandCenter_, script, valueVarArray, ref values_, withEcho_) )
            {
                return false;
            }

            return true;
        }

        //=================================================================================================
        //   Retrieves the values of each variable when running a given script
        //
        // Parameter : commandCenter_ : command center
        // Parameter : script_        : script to execute
        // Parameter : varNames_      : name of the variables to retrieve
        // Parameter : values_        : retrieved values of each variable
        // Parameter : withEcho_      : indicates whether the command should be visible in the application's Command History
        //
        // Return : true if successful, false otherwise
        // -------------------------------------------------------------------------------------------------
        private bool RunScriptAndReadValueArrays(
            IMInspectLib.IIMCommandCenter commandCenter_,
            String script_,
            System.Collections.Generic.List<String> varNames_,
            ref System.Collections.Generic.List<System.Collections.Generic.List<String>> values_,
            bool withEcho_)
        {
            if( withEcho_)
            {
                commandCenter_.CommandExecute("EchoOn");
                commandCenter_.CommandExecute("RecordableOn");
            }
            else
            {
                commandCenter_.CommandExecute("EchoOff");
                commandCenter_.CommandExecute("RecordableOff");

            }

            // Run the script
            int returnValue = commandCenter_.ScriptExecuteFromBuffer(script_, "");
            if (commandCenter_.ReturnValueIsSuccess(returnValue) == 0) return false;

            // Retrieves the values of each variable
            for ( int i = 0; i < varNames_.Count ; i++)
            {
                int nbValues;
                System.Collections.Generic.List<String> values = new System.Collections.Generic.List<String>();

                nbValues = commandCenter_.ScriptVariableGetNbValues(varNames_[i]);

                for( int j = 0; j < nbValues; j++)
                {
                    String value;
                    returnValue = commandCenter_.ScriptVariableGetValueAsString(varNames_[i], j+1, out value);
                    if (commandCenter_.ReturnValueIsSuccess(returnValue) == 1)
                    {
                        values.Add(value);
                    }
                }
                values_.Add(values);
            }
            return true;
        }
    }
}